✨[FEAT] #13: 유저 피드 기능 구현

kimeunsom opened the pull request:
## #⃣ 연관된 이슈

> - close #13 

## 📝 작업 내용

1. 유저 기본 정보 조회 기능( 본인의 프로필사진, 닉네임, @아이디, 팔로잉수, 팔로워수, 한줄소개, 최근게시글 여부를 반환하는 기능 )
2. 유저가 만든 쇼츠를 리스트 조회하는 기능
3. 유저가 찜한 쇼츠를 리스트 조회하는 기능
4. 유저가 읽은 책 리스트를 조회하는 기능

### 📸 스크린샷 (선택)

## 💬 리뷰 요구사항(선택)

Refactor: 프로필 띠 기능 쿼리 수정
Refactor: 다른 유저 프로필보는 컨트롤러 수정 

메이님이 수정 요청하신 사항 2가지 위의 2가지 커밋에 각각 담았습니다 다시 리뷰 부탁드려요 :)
일단 api는 잘 돌아가는 것 같아욥 자세한 수정 사항은 커밋문에 줄글로 설명해두었어요 !

+) Refactor: 다른 유저 조회 시, 팔로우 여부도 반환
다른 유저 프로필 볼 때, 팔로우 여부 boolean값도 리턴되어야했는데 빼먹어서 새로 추가했습니다 !
커밋문에 수정해두었듯이, 마이페이지 / 다른 유저 피드 조회용 dto를 분리하고, findOne 함수도 분리했습니다 !

kimeunsom committed:
FEAT: 다른 유저 프로필 기능 추가

kimeunsom committed:
FEAT: users 병합 충돌  해결

kimeunsom committed:
FEAT: 다른 유저 프로필 조회 컨트롤러 수정중

kimeunsom committed:
Merge branch 'dev' of https://github.com/UMCreadme/Read_Me_Node.js into feat/13

kimeunsom committed:
Merge branch 'dev' of https://github.com/UMCreadme/Read_Me_Node.js into feat/13

kimeunsom committed:
FEAT: 프로필 띠 기능 추가
- 24시간 이내 게시물 게시 여부를 유저 정보 반환값에 추가

SuHyeon00 commented on src/users/users.route.js:
```diff
@@ -20,5 +19,11 @@ userRouter.get("/my/books", asyncHandler(getUserBooks));
+userRouter.get('/:userId', asyncHandler(getOtherUserInfo));
+
+// 다른 유저의 피드 조회 (로그인 필요 X)
+userRouter.get('/:userId/shorts', asyncHandler(getUserShorts));
+userRouter.get('/:userId/likes', asyncHandler(getUserLikeShorts));
+userRouter.get('/:userId/books', asyncHandler(getUserBooks));
```
현재 본인 유저의 피드 조회 부분의 컨트롤러로 연결되어 있는 것 같은데 이 부분도 유저 정보 조회 구성과 비슷하게 컨트롤러는 회원 / 비회원을 구분짓고 서비스는 같은 로직을 쓰되 전달되는 값이 userId 값이 되도록 수정 부탁드려요!

kimeunsom replied:
forGuest 함수들 각각 만들어 수정했습니다 !

SuHyeon00 commented on src/users/users.dao.js:
```diff
@@ -67,6 +68,24 @@ export const findFollowerNumByUserId = async (userId) => {
+    
+    try {
+        const [recent] = await conn.query(getLatestPostTime, [userId]);
+        const latestPostTime = recent.length > 0 ? recent[0].created_at : null;
+
+        // 프로필 띠 기능 - 현재 시간과 게시물 시간을 비교하여 24시간 이내인지 확인
+        const hasRecentPost = latestPostTime ? (new Date() - new Date(latestPostTime)) < 24 * 60 * 60 * 1000 : false;
```
이걸 쿼리단에서 바로 확인할 수 있는 방법이 있어요!! 아래와 같이 DATE_SUB 함수로 현재 시각으로부터 24시간 전 이후에 있는 쇼츠만 갖고올 수 있습니다!

```sql
SELECT shorts_id
FROM SHORTS
WHERE user_id = 3 AND created_at >= DATE_SUB(NOW(), INTERVAL 24 HOUR);
```

kimeunsom replied:
감사합니닷 수정했어요 ! :)

SuHyeon00 reviewed (changes requested):
컨트롤러 회원 / 비회원 나눠주는 부분 수정 부탁드립니다! 아마 찜한 쇼츠쪽에서 오류 뜨는 것도 이쪽과 관련된 문제일 수 있어서 이 부분 수정 후에도 오류가 뜬다면 다시 호출해주세요!

kimeunsom committed:
Merge branch 'dev' of https://github.com/UMCreadme/Read_Me_Node.js into feat/13

kimeunsom committed:
Refactor: 프로필 띠 기능 쿼리 수정
- 24시간 이내 게시물 수를 가져오도록 sql 쿼리 수정. 개수가 1개 이상이면 true로 반환
- (기존) 가장 최근 게시물 가져와서 24시간 이내인지 확인하는 방식

kimeunsom committed:
Merge branch 'dev' of https://github.com/UMCreadme/Read_Me_Node.js into feat/13

kimeunsom committed:
Merge branch 'dev' of https://github.com/UMCreadme/Read_Me_Node.js into feat/13

kimeunsom committed:
Refactor: 다른 유저 프로필보는 컨트롤러 수정
- shorts.sql.js : 테이블명 소문자 에러 수정
- 컨트롤러 마이페이지용, 비회원사용가능용(forGuest)  분리
- req.body -> req.body.id 로 객체형에서 단일값으로 추출. 그리고 모든 ID값 인트형으로 통일

urimida commented on src/users/users.dto.js:
```diff
@@ -1,14 +1,16 @@
         "followingNum" : followingNum,
-        "profileImg" : userData.image_url }
+        "profileImg" : userData.image_url,
+        "isRecentPost": isRecentPost 
```
{
    "isSuccess": true,
    "code": 2000,
    "message": "SUCCESS!",
    "result": {
        "userId": 1,
        "nickname": "ossu",
        "account": "ossu-dev",
        "comment": "HIHI",
        "followerNum": 1,
        "followingNum": 4,
        "profileImg": "https://readme-image.s3.ap-northeast-2.amazonaws.com/profile/1722006846344_ossu-profile.jpeg",
        "isRecentPost": false
    }
}

여기서 명세서와는 다르게 isRecentPost부분이 있어서 명세서 수정이 필요할 것 같아요!

kimeunsom replied:
아 넵 프로필 띠 기능 추가되었는데, api 명세 수정을 깜빡했네요 !! 감사합니다 :)

urimida commented on src/users/users.controller.js:
```diff
@@ -54,7 +90,22 @@ export const getUserBooks = async(req, res, next)=> {
+    res.send(response(status.SUCCESS, result, pageInfo(page, result.length, hasNext)))
+}
+
+// 유저가 읽은 책 리스트 조회 (비회원)
```
{
            "bookImage": "https://image.aladin.co.kr/product/17531/63/cover500/k062534304_1.jpg",
            "bookTitle": "돌이킬 수 있는",
            "bookAuthor": "문목하",
            "bookTranslator": null
        },
        {
            "bookImage": "https://image.aladin.co.kr/product/34312/24/cover500/8932924376_1.jpg",
            "bookTitle": "고요의 바다에서",
            "bookAuthor": "에밀리 세인트존 맨델",
            "bookTranslator": "강동혁"
        },
        {
            "bookImage": "https://image.aladin.co.kr/product/19359/16/cover500/s152835852_1.jpg",
            "bookTitle": "우리가 빛의 속도로 갈 수 없다면",
            "bookAuthor": "김초엽",
            "bookTranslator": ""
        },
        {
            "bookImage": "https://image.aladin.co.kr/product/34360/88/cover500/k182932901_1.jpg",
            "bookTitle": "디자인구구단",
            "bookAuthor": "에이핫",
            "bookTranslator": ""
        }
    ]
}
이렇게 나오는데 어떤 건 옮긴이가 null이고 어떤 건 아예 "" 상태로 비워져 있어서 확인부탁드려요!!

kimeunsom replied:
값이 없는 경우엔 아예 result에서 번역자가 제외되도록 수정했습니다 ! 감사해요 :)

urimida reviewed (changes requested):
구현하시느라 너무 수고 많으셨어요! 리뷰 남긴 건 의문들인데 답변 읽어보고 승인해드리겠습니다!!! 일찍 주무세요....

kimeunsom committed:
Refactor: 다른 유저 조회 시, 팔로우 여부도 반환
- 마이페이지를 위한 유저 정보 조회와 달리, 다른 유저 피드를 볼 때는 팔로우 여부도 반환해주어야함
그래서 DTO를 분리하고, findOne도 분리 (user.service.js)

kimeunsom committed:
Refactor:  충돌 해결중

kimeunsom committed:
Merge branch 'dev' of https://github.com/UMCreadme/Read_Me_Node.js into feat/13

kimeunsom committed:
Merge branch 'dev' of https://github.com/UMCreadme/Read_Me_Node.js into feat/13

kimeunsom committed:
Refactor: 카카오 로그인 기능 추가

kimeunsom committed:
Refactor: 충돌 해결

kimeunsom committed:
Merge branch 'dev' of https://github.com/UMCreadme/Read_Me_Node.js into feat/13

kimeunsom committed:
Refactor: 충돌 해결

kimeunsom committed:
refactor: 충돌 수정

SuHyeon00 committed:
🐛 Fix(#13): 문법 오류 수정

SuHyeon00 committed:
🐛 Fix(#13): import문 중복 삽입 수정

SuHyeon00 reviewed (approved):
기능 잘 돌아가는 것 같습니다! 수고하셨어요 👍

kimeunsom committed:
bug: 충돌 오타 삭제..

urimida reviewed (approved):
확인했어요!! 너무 수고하셨어요~~ 코튼님최고
